File: solutions/backtracking/typescript/0046-permutations.ts
/**
 * ### METADATA:
 * **Techniques**: Backtracking, Recursion
 * **Data Structures**: Array, List
 * **Time Complexity**: O(n × n!)
 * **Space Complexity**: O(n)
 *
 * ### INTUITION:
 * The key insight is that generate all permutations by systematically trying each unused element at each position. Use backtracking to explore all possibilities while maintaining state through choices and un-choices.
 *
 * ### APPROACH:
 * 1. **Initialize result list**: Create an empty list to store all permutations
 * 2. **Define recursive backtracking function**: Create a helper function that builds permutations by maintaining a current partial permutation
 * 3. **Base case check**: If the current permutation length equals the input array length, we have a complete permutation - add a copy to results
 * 4. **Iterate through all elements**: For each element in the original array, check if it's not already in the current permutation
 * 5. **Make choice and recurse**: Add the element to current permutation, recursively call backtrack to continue building
 * 6. **Backtrack**: Remove the last added element (undo choice) to try other possibilities
 * 7. **Return all permutations**: After exploring all branches, return the complete list of permutations
 *
 * ### WHY THIS WORKS:
 * - Each permutation uses every element exactly once
 * - Backtracking ensures we explore all n! permutations
 * - Checking "not in current" ensures no duplicates within a permutation
 * - Systematic exploration guarantees all permutations are found
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * nums = [1, 2, 3]
 * ```
 *
 * Step 1: backtrack([])
 * Try 1: current = [1]
 * Try 2: current = [1,2]
 * Try 3: current = [1,2,3] ✓ Complete! Add to result
 * Remove 3: current = [1,2]
 * Remove 2: current = [1]
 * Try 3: current = [1,3]
 * Try 2: current = [1,3,2] ✓ Complete! Add to result
 * Remove 2: current = [1,3]
 * Remove 3: current = [1]
 * Remove 1: current = []
 * Step 2: Try 2: current = [2]
 * Try 1: current = [2,1]
 * Try 3: current = [2,1,3] ✓ Complete! Add to result
 * Try 3: current = [2,3]
 * Try 1: current = [2,3,1] ✓ Complete! Add to result
 * Step 3: Try 3: current = [3]
 * Try 1: current = [3,1]
 * Try 2: current = [3,1,2] ✓ Complete! Add to result
 * Try 2: current = [3,2]
 * Try 1: current = [3,2,1] ✓ Complete! Add to result
 *
 * Output:
 * ```
 * [[1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], [3,2,1]]
 * ```
 *
 * ### TIME COMPLEXITY:
 * **O(n × n!)** - where n is the number of elements in the input array. There are n! (n factorial) possible permutations of n distinct elements. For each permutation, we perform **O(n)** work: checking membership in the current permutation (**O(n)** for list lookup), adding elements (**O(1)**), and copying the complete permutation to the result (**O(n)**). Therefore, the total time complexity is **O(n!)** permutations × **O(n)** work per permutation = **O(n × n!)**. Note that while the first call explores n choices, the second explores n-1, then n-2, etc., giving us n × (n-1) × (n-2) × ... × 1 = n! total paths.
 *
 * ### SPACE COMPLEXITY:
 * **O(n)** - where n is the number of elements in the input array. The recursion call stack can go as deep as n levels (one level for each element we add to the current permutation). At each level, we maintain a current permutation list that grows from size 0 to size n. The maximum stack depth is n when we have a complete permutation. Additionally, the current permutation list takes **O(n)** space. The result list storing all permutations is not counted toward space complexity as it's required output. Total auxiliary space: **O(n)** for recursion stack + **O(n)** for current permutation = **O(n)**.
 *
 * ### EDGE CASES:
 * - **Empty array**: Return [[]] (empty permutation)
 * - **Single element**: Return [[element]]
 * - **Two elements**: Return both orderings [[a,b], [b,a]]
 * - **Duplicate elements (in variant)**: Use frequency counter to avoid duplicates
 * - **Large arrays**: n! permutations, factorial growth
 *
 *
 */

class Solution {
  /**
   * Approach: Backtracking
   * Time Complexity: O(n * n!)
   * Space Complexity: O(n) for recursion
   */
  permute(nums: number[]): number[][] {
    const result: number[][] = [];

    const backtrack = (current: number[]): void => {
      if (current.length === nums.length) {
        result.push([...current]);
        return;
      }

      for (const num of nums) {
        if (!current.includes(num)) {
          current.push(num);
          backtrack(current);
          current.pop();
        }
      }
    };

    backtrack([]);
    return result;
  }
}

if (typeof module !== "undefined" && module.exports) {
  module.exports = Solution;
}

function runTests(): void {
  const solution = new Solution();

  const result1 = solution.permute([1, 2, 3]);
  console.log(`Test 1: ${result1.length === 6 ? "PASS" : "FAIL"}`);

  const result2 = solution.permute([0, 1]);
  console.log(`Test 2: ${result2.length === 2 ? "PASS" : "FAIL"}`);

  const result3 = solution.permute([1]);
  console.log(`Test 3: ${result3.length === 1 ? "PASS" : "FAIL"}`);

  console.log("\nAll test cases completed!");
}

if (typeof require !== "undefined" && require.main === module) {
  runTests();
}

export default Solution;

File: solutions/backtracking/python/0046-permutations.py
"""
### INTUITION:
The key insight is that generate all permutations by systematically trying each unused element at each position. Use backtracking to explore all possibilities while maintaining state through choices and un-choices.

### APPROACH:
1. **Initialize result list**: Create an empty list to store all permutations
2. **Define recursive backtracking function**: Create a helper function that builds permutations by maintaining a current partial permutation
3. **Base case check**: If the current permutation length equals the input array length, we have a complete permutation - add a copy to results
4. **Iterate through all elements**: For each element in the original array, check if it's not already in the current permutation
5. **Make choice and recurse**: Add the element to current permutation, recursively call backtrack to continue building
6. **Backtrack**: Remove the last added element (undo choice) to try other possibilities
7. **Return all permutations**: After exploring all branches, return the complete list of permutations

### WHY THIS WORKS:
- Each permutation uses every element exactly once
- Backtracking ensures we explore all n! permutations
- Checking "not in current" ensures no duplicates within a permutation
- Systematic exploration guarantees all permutations are found

### EXAMPLE WALKTHROUGH:
Input:
```
nums = [1, 2, 3]
```

Step 1: backtrack([])
Try 1: current = [1]
Try 2: current = [1,2]
Try 3: current = [1,2,3] ✓ Complete! Add to result
Remove 3: current = [1,2]
Remove 2: current = [1]
Try 3: current = [1,3]
Try 2: current = [1,3,2] ✓ Complete! Add to result
Remove 2: current = [1,3]
Remove 3: current = [1]
Remove 1: current = []
Step 2: Try 2: current = [2]
Try 1: current = [2,1]
Try 3: current = [2,1,3] ✓ Complete! Add to result
Try 3: current = [2,3]
Try 1: current = [2,3,1] ✓ Complete! Add to result
Step 3: Try 3: current = [3]
Try 1: current = [3,1]
Try 2: current = [3,1,2] ✓ Complete! Add to result
Try 2: current = [3,2]
Try 1: current = [3,2,1] ✓ Complete! Add to result

Output:
```
[[1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], [3,2,1]]
```

### TIME COMPLEXITY:
**O(n × n!)** - where n is the number of elements in the input array. There are n! (n factorial) possible permutations of n distinct elements. For each permutation, we perform **O(n)** work: checking membership in the current permutation (**O(n)** for list lookup), adding elements (**O(1)**), and copying the complete permutation to the result (**O(n)**). Therefore, the total time complexity is **O(n!)** permutations × **O(n)** work per permutation = **O(n × n!)**. Note that while the first call explores n choices, the second explores n-1, then n-2, etc., giving us n × (n-1) × (n-2) × ... × 1 = n! total paths.

### SPACE COMPLEXITY:
**O(n)** - where n is the number of elements in the input array. The recursion call stack can go as deep as n levels (one level for each element we add to the current permutation). At each level, we maintain a current permutation list that grows from size 0 to size n. The maximum stack depth is n when we have a complete permutation. Additionally, the current permutation list takes **O(n)** space. The result list storing all permutations is not counted toward space complexity as it's required output. Total auxiliary space: **O(n)** for recursion stack + **O(n)** for current permutation = **O(n)**.

### EDGE CASES:
- **Empty array**: Return [[]] (empty permutation)
- **Single element**: Return [[element]]
- **Two elements**: Return both orderings [[a,b], [b,a]]
- **Duplicate elements (in variant)**: Use frequency counter to avoid duplicates
- **Large arrays**: n! permutations, factorial growth

</details>

"""

from collections import Counter
from typing import Any


class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        """
        Approach: Backtracking
        Time Complexity: O(n * n!)
        Space Complexity: O(n) for recursion
        """
        result: list[Any] = []

        def backtrack(current: Any) -> Any:
            if len(current) == len(nums):
                result.append(current[:])
                return

            for num in nums:
                if num not in current:
                    current.append(num)
                    backtrack(current)
                    current.pop()

        backtrack([])
        return result

    def permuteSwap(self, nums: list[int]) -> list[list[int]]:
        """
        Approach: Backtracking with swapping
        Time Complexity: O(n * n!)
        Space Complexity: O(n)
        """
        result: list[list[int]] = []

        def backtrack(first: Any = 0) -> Any:
            if first == len(nums):
                result.append(nums[:])
                return

            for i in range(first, len(nums)):
                nums[first], nums[i] = nums[i], nums[first]
                backtrack(first + 1)
                nums[first], nums[i] = nums[i], nums[first]

        backtrack()
        return result


"""
47. Permutations II
Given a collection of numbers, nums, that might contain duplicates, return all
possible unique permutations in any order.

Example:
Input: nums = [1,1,2]
Output: [[1,1,2],[1,2,1],[2,1,1]]
"""


class SolutionUnique:
    def permuteUnique(self, nums: list[int]) -> list[list[int]]:
        """
        Approach: Backtracking with frequency map
        Time Complexity: O(n * n!)
        Space Complexity: O(n)
        """
        result: list[Any] = []
        counter = Counter(nums)

        def backtrack(current: Any) -> Any:
            if len(current) == len(nums):
                result.append(current[:])
                return

            for num in counter:
                if counter[num] > 0:
                    current.append(num)
                    counter[num] -= 1
                    backtrack(current)
                    current.pop()
                    counter[num] += 1

        backtrack([])
        return result


"""
31. Next Permutation
A permutation of an array of integers is an arrangement of its members into a
sequence or linear order.

Implement next permutation, which rearranges numbers into the lexicographically
next greater permutation of numbers.

Example:
Input: nums = [1,2,3]
Output: [1,3,2]
"""


class SolutionNext:
    def nextPermutation(self, nums: list[int]) -> None:
        """
        Approach: Find pivot and swap
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        n = len(nums)

        # Find the pivot (first decreasing element from right)
        pivot = n - 2
        while pivot >= 0 and nums[pivot] >= nums[pivot + 1]:
            pivot -= 1

        if pivot >= 0:
            # Find the smallest element greater than pivot
            successor = n - 1
            while nums[successor] <= nums[pivot]:
                successor -= 1

            # Swap
            nums[pivot], nums[successor] = nums[successor], nums[pivot]

        # Reverse the suffix
        left, right = pivot + 1, n - 1
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1


# Test cases
if __name__ == "__main__":
    # Test Permutations
    solution = Solution()

    print("Permutations:")
    test_cases = [[1, 2, 3], [0, 1], [1]]
    for nums in test_cases:
        result = solution.permute(nums)
        print(f"Input: nums")
        print(f"Permutations: result\n")

    # Test Permutations with Duplicates
    solution_unique = SolutionUnique()

    print("Unique Permutations:")
    test_cases_dup = [[1, 1, 2], [1, 2, 3], [2, 2, 1, 1]]
    for nums in test_cases_dup:
        result = solution_unique.permuteUnique(nums)
        print(f"Input: nums")
        print(f"Unique Permutations: result\n")

    # Test Next Permutation
    solution_next = SolutionNext()

    print("Next Permutation:")
    test_cases_next = [[1, 2, 3], [3, 2, 1], [1, 1, 5], [1]]
    for nums in test_cases_next:
        original = nums.copy()
        solution_next.nextPermutation(nums)
        print(f"Input: {original} -> Next: nums")
